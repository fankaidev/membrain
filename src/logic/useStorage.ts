import { useEffect, useRef, useState } from "react";

export function useStorage<T>(
  areaName: "sync" | "local",
  keyInArea: string,
  defaultValue: T
): [T, (value: T) => void] {
  const key = `${areaName}:${keyInArea}`;
  const [value, setValue] = useState<T>(defaultValue);
  const valueRef = useRef(defaultValue);
  const area = areaName === "sync" ? chrome.storage.sync : chrome.storage.local;

  useEffect(() => {
    area.get([key], (result) => {
      if (result[key] && result[key] !== JSON.stringify(valueRef.current)) {
        console.debug(`init from storage ${key}=${result[key]}`);
        valueRef.current = JSON.parse(result[key]);
        setValue(valueRef.current);
      }
    });
  }, []);

  useEffect(() => {
    if (value !== valueRef.current) {
      valueRef.current = value;
      // console.debug(`set ${key} value`);
      area.set({ [key]: JSON.stringify(value) });
    }
  }, [value]);

  return [value, (value: T) => setValue(value)];
}

export function useSyncStorage<T>(key: string, defaultValue: T): [T, (value: T) => void] {
  return useStorage("sync", key, defaultValue);
}

export function useLocalStorage<T>(key: string, defaultValue: T): [T, (value: T) => void] {
  return useStorage("local", key, defaultValue);
}

export async function getFromStorage<T>(areaName: string, keyInArea: string, defaultValue: T) {
  const area = areaName === "sync" ? chrome.storage.sync : chrome.storage.local;
  const key = `${areaName}:${keyInArea}`;
  const data = await area.get([key]);
  return data[key] ? (JSON.parse(data[key]) as T) : defaultValue;
}

export async function saveToStorage<T>(areaName: string, keyInArea: string, value: T) {
  const area = areaName === "sync" ? chrome.storage.sync : chrome.storage.local;
  const key = `${areaName}:${keyInArea}`;
  await area.set({ [key]: JSON.stringify(value) });
}
