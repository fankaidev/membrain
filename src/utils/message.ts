import { uniqueId } from "lodash";
import { v4 as uuidv4 } from "uuid";

export class Message {
  role: string;
  content: string;
  model: string;
  temperature: number;

  constructor(role: string, content: string, model: string = "", temperature: number = -1) {
    this.role = role;
    this.content = content;
    this.model = model;
    this.temperature = temperature;
  }
}

export type ReferenceType = "text" | "webpage";

export class Reference {
  id: string;
  type: ReferenceType;
  title: string;
  url: string;
  content: string;

  constructor(type: ReferenceType, title: string, url: string, content: string) {
    this.type = type;
    this.title = title;
    this.url = url;
    this.content = content;
    if (type === "webpage") {
      this.id = `${type}:${url}`;
    } else {
      this.id = `${type}:${title}`;
    }
  }
}

export type ChatReferenceType = "all" | "page" | "selection";

export class ChatTask {
  id: string;
  prompt: string;
  reference_type: ChatReferenceType;

  constructor(prompt: string, reference_type: ChatReferenceType) {
    this.prompt = prompt;
    this.reference_type = reference_type;
    this.id = new Date().toISOString() + "_" + uniqueId();
  }
}

export class PromptTemplate {
  id: string;
  name: string;
  prompt: string;
  reference_type: ChatReferenceType;

  constructor(name: string, prompt: string, reference_type: ChatReferenceType) {
    this.id = uuidv4();
    this.name = name;
    this.prompt = prompt;
    this.reference_type = reference_type;
  }
}

export const CHAT_STATUS_PROCESSING = "__processing__";
export const CHAT_STATUS_EMPTY = "__empty__";
