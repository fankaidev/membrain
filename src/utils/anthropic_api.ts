import Anthropic from "@anthropic-ai/sdk";
import { Model } from "./config";
import { Message } from "./message";

const prepareChatData = (
  query_messages: Message[],
  model: string,
  temperature: number,
  max_tokens: number
) => {
  const system_msg = query_messages[0].content;
  const messages = query_messages
    .slice(1)
    .filter((message) => message.content && message.content.trim() != "<error>")
    .map((message) => ({
      role: message.role as "user" | "assistant",
      content: message.content,
    }));

  const data = {
    model,
    system: system_msg,
    messages,
    temperature,
    max_tokens,
  };
  return data;
};

export const callClaude = async (
  apiKey: string,
  model: Model,
  messages: Message[],
  onContent: (_: string) => void,
  onFinish: (_?: string) => void
) => {
  const anthropic = new Anthropic({ apiKey });
  const data = prepareChatData(messages, model.name, 0.3, model.maxTokens);
  await anthropic.messages
    .stream(data)
    .on("text", (text) => {
      onContent(text);
    })
    .on("error", (error) => {
      onFinish(error.message);
    })
    .on("end", () => {
      onFinish();
    });
};
