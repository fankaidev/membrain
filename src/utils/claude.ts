import Anthropic from "@anthropic-ai/sdk";
import { MessageParam } from "@anthropic-ai/sdk/resources/index.mjs";
import { Message } from "./message";

function prepareChatData(
  query_messages: Message[],
  model: string,
  temperature: number,
  max_tokens: number
) {
  const messages: MessageParam[] = [];
  for (const message of query_messages) {
    if (message.content && message.content.trim() != "<error>") {
      messages.push({
        role: message.role === "assistant" ? "assistant" : "user",
        content: message.content,
      });
    }
  }
  const data = {
    model,
    messages,
    temperature,
    max_tokens,
  };
  return data;
}

export const callClaude = async (
  apiKey: string,
  messages: Message[],
  onContent: (_: string) => void,
  onFinish: (_?: string) => void
) => {
  const anthropic = new Anthropic({ apiKey });
  if (messages.length > 1) {
    const first = new Message("user", messages[0].content + "\n\n" + messages[1].content);
    messages = [first, ...messages.slice(2)];
  }
  const data = prepareChatData(messages, "claude-3-sonnet-20240229", 0.3, 2048);
  await anthropic.messages
    .stream(data)
    .on("text", (text) => {
      onContent(text);
    })
    .on("error", (error) => {
      onFinish(error.message);
    });
  onFinish();
};
