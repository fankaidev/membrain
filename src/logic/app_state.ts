import { create } from "zustand";
import { getLocaleMessage } from "../utils/locale";
import { getFromStorage, saveToStorage } from "./chrome_storage";

export type AppState = {
  UILanguage: string;
  chatLanguage: string;
  setUILanguage: (val: string) => void;
  setChatLanguage: (val: string) => void;
  loadAppState: () => Promise<void>;
  displayText: (text: string) => string;
};

export const useAppState = create<AppState>((set, get) => ({
  UILanguage: "en",
  chatLanguage: "English",
  setUILanguage: (val: string) => {
    set({ UILanguage: val });
    saveToStorage("sync", "UILanguage", val);
  },
  setChatLanguage: (val: string) => {
    set({ chatLanguage: val });
    saveToStorage("sync", "chatLanguage", val);
  },
  loadAppState: async () => {
    const UILanguage = await getFromStorage("sync", "UILanguage", "en");
    const chatLanguage = await getFromStorage("sync", "chatLanguage", "English");
    set({ UILanguage, chatLanguage });
  },
  displayText: (text: string) => {
    return getLocaleMessage(get().UILanguage, text);
  },
}));
