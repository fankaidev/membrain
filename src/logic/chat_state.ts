import { create } from "zustand";
import {
  Model,
  ModelAndProvider,
  ModelProvider,
  ProviderConfig,
  SYSTEM_MODELS,
  SYSTEM_PROVIDERS,
} from "../utils/config";
import { CHAT_STATUS_EMPTY, ChatTask, Message, PromptTemplate } from "../utils/message";
import { getFromStorage, saveToStorage } from "./chrome_storage";

export type ChatState = {
  chatStatus: string;
  chatTask: ChatTask | null;
  temperature: number;
  promptTemplates: PromptTemplate[];
  customModels: Model[];
  customProviders: ModelProvider[];
  providerConfigs: Record<string, ProviderConfig>;
  selectedModel: ModelAndProvider | null;
  history: Message[];
  loaded: boolean;
  setChatStatus: (val: string) => void;
  setChatTask: (val: ChatTask | null) => void;
  setTemperature: (val: number) => void;
  setPromptTemplates: (val: PromptTemplate[]) => void;
  setCustomModels: (val: Model[]) => void;
  setCustomProviders: (val: ModelProvider[]) => void;
  setProviderConfigs: (val: Record<string, ProviderConfig>) => void;
  setSelectedModel: (val: ModelAndProvider | null) => void;
  setHistory: (val: Message[]) => void;
  getAllModels: () => Model[];
  getAllProviders: () => ModelProvider[];
  getEnabledModels(): ModelAndProvider[];
  getCurrentModel: () => ModelAndProvider | null;
  loadChatState: () => Promise<void>;
  clearChatSession: () => void;
};

export const useChatState = create<ChatState>((set, get) => ({
  chatStatus: CHAT_STATUS_EMPTY,
  chatTask: null,
  temperature: 0.3,
  promptTemplates: [],
  customModels: [],
  customProviders: [],
  providerConfigs: {},
  selectedModel: null,
  history: [],
  loaded: false,
  setChatStatus: (val: string) => {
    console.debug("setChatStatus", val);
    set({ chatStatus: val });
  },
  setChatTask: (val: ChatTask | null) => {
    set({ chatTask: val });
  },
  setTemperature: (val: number) => {
    set({ temperature: val });
    saveToStorage("sync", "temperature", val);
  },
  setPromptTemplates: (val: PromptTemplate[]) => {
    set({ promptTemplates: val });
    saveToStorage("sync", "promptTemplates", val);
  },
  setCustomModels: (val: Model[]) => {
    set({ customModels: val });
    saveToStorage("sync", "customModels", val);
  },
  setCustomProviders: (val: ModelProvider[]) => {
    set({ customProviders: val });
    saveToStorage("sync", "customProviders", val);
  },
  setProviderConfigs: (val: Record<string, ProviderConfig>) => {
    set({ providerConfigs: val });
    saveToStorage("sync", "providerConfigs", val);
  },
  setSelectedModel: (val: ModelAndProvider | null) => {
    set({ selectedModel: val });
    console.debug("setSelectedModel", val?.model.id);
    saveToStorage("sync", "selectedModelId", val?.model.id);
  },
  setHistory: (val: Message[]) => {
    set({ history: val });
    saveToStorage("local", "history", val);
  },
  getAllModels: () => {
    return [...SYSTEM_MODELS, ...get().customModels];
  },
  getAllProviders: () => {
    return [...SYSTEM_PROVIDERS, ...get().customProviders];
  },
  getEnabledModels: () => {
    const { getAllProviders, providerConfigs, getAllModels } = get();
    return getAllProviders()
      .map((p) => [p, providerConfigs[p.id]] as [ModelProvider, ProviderConfig])
      .filter(([p, c]) => c && c.enabled)
      .flatMap(([p, c]) =>
        getAllModels()
          .filter((m) => m.providerId === p.id && c.enabledModels.includes(m.name))
          .map((m) => new ModelAndProvider(m, p)),
      );
  },
  getCurrentModel: () => {
    const { getEnabledModels, selectedModel } = get();
    const enabledModels = getEnabledModels();

    const isSelectedModelValid = (): boolean => {
      if (selectedModel === null) {
        return false;
      }
      const { model, provider } = selectedModel;
      return !!enabledModels.find((mp) => mp.model.id == model.id && mp.provider.id == provider.id);
    };

    if (enabledModels.length === 0) {
      return null;
    }
    if (enabledModels.length > 0 && !isSelectedModelValid()) {
      set({ selectedModel: enabledModels[0] });
      return enabledModels[0];
    }
    return selectedModel;
  },
  loadChatState: async () => {
    const temperature = await getFromStorage("sync", "temperature", 0.3);
    const promptTemplates = await getFromStorage("sync", "promptTemplates", []);
    const customModels = await getFromStorage("sync", "customModels", []);
    const customProviders = await getFromStorage("sync", "customProviders", []);
    const providerConfigs = await getFromStorage("sync", "providerConfigs", {});
    const history = await getFromStorage("local", "history", []);
    set({ temperature, promptTemplates, customModels, customProviders, providerConfigs, history });

    const selectedModelId = await getFromStorage("sync", "selectedModelId", null);
    const selectedModel = get()
      .getEnabledModels()
      .find((m) => m.model.id === selectedModelId);
    if (selectedModel) {
      set({ selectedModel });
    }

    console.debug("loaded enabled models:", get().getEnabledModels());
    console.debug("loaded selectedModel", selectedModelId, selectedModel);
    set({ loaded: true });
  },
  clearChatSession: () => {
    set({
      chatStatus: CHAT_STATUS_EMPTY,
      chatTask: null,
      history: [],
    });
  },
}));
