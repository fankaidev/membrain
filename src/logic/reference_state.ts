import { create } from "zustand";
import { Reference } from "../utils/message";
import { getCurrentPageRef, getCurrentSelectionRef } from "../utils/page_content";
import { getFromStorage, saveToStorage } from "./useStorage";

export type ChatReferenceStore = {
  references: Reference[];
  addPageRef: () => Promise<Reference | null>;
  addSelectionRef: () => Promise<Reference | null>;
  removeRef: (id: string) => void;
  clear: () => void;
  load: () => Promise<void>;
};

export const useReferenceStore = create<ChatReferenceStore>((set) => ({
  references: [],

  addPageRef: async () => {
    const pageRef = await getCurrentPageRef();
    if (!pageRef) {
      return null;
    }

    set((state) => {
      if (!state.references.find((r) => r.type === "webpage" && r.url === pageRef.url)) {
        const result = [...state.references, pageRef];
        saveToStorage("local", "references", result);
        return { references: result };
      } else {
        console.debug("skip adding existing reference");
        return state;
      }
    });
    return pageRef;
  },

  addSelectionRef: async () => {
    const selectionRef = await getCurrentSelectionRef();
    if (!selectionRef) {
      return null;
    }

    set((state) => {
      const result = [...state.references, selectionRef];
      saveToStorage("local", "references", result);
      return { references: result };
    });
    return selectionRef;
  },

  removeRef: (id: string) => {
    set((state) => {
      const result = state.references.filter((r) => r.id !== id);
      saveToStorage("local", "references", result);
      return { references: result };
    });
  },

  clear: () => {
    set({ references: [] });
    saveToStorage("local", "references", []);
  },

  load: async () => {
    const references = await getFromStorage("local", "references", []);
    set({ references });
  },
}));
