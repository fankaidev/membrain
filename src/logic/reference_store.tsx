import { Reference } from "../utils/message";
import { getCurrentPageRef, getCurrentSelectionRef } from "../utils/page_content";

import { create } from "zustand";

export type ChatReferenceStore = {
  references: Reference[];
  addPageRef: () => Promise<Reference | null>;
  addSelectionRef: () => Promise<Reference | null>;
  removeReference: (id: string) => void;
  clearReferences: () => void;
};

export const useChatReferenceStore = create<ChatReferenceStore>((set) => ({
  references: [],

  addPageRef: async () => {
    const pageRef = await getCurrentPageRef();
    if (!pageRef) {
      return null;
    }

    set((state) => {
      if (!state.references.find((r) => r.type === "webpage" && r.url === pageRef.url)) {
        return { references: [...state.references, pageRef] };
      } else {
        console.debug("skip adding existing reference");
        return state;
      }
    });
    return pageRef;
  },

  addSelectionRef: async () => {
    const selectionRef = await getCurrentSelectionRef();
    if (selectionRef) {
      set((state) => ({
        references: [...state.references, selectionRef],
      }));
      return selectionRef;
    } else {
      return null;
    }
  },

  removeReference: (id: string) => {
    set((state) => ({
      references: state.references.filter((r) => r.id !== id),
    }));
  },

  clearReferences: () => {
    set({ references: [] });
  },
}));
