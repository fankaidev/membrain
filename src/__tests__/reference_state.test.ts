const storageMock = (() => {
  let storage: { [key: string]: string } = {};

  return {
    set: jest.fn((items, callback = () => {}) => {
      return new Promise<void>((resolve) => {
        console.log("set items", items);
        Object.keys(items).forEach((key) => {
          storage[key] = items[key];
        });
        callback();
        resolve();
      });
    }),
    get: jest.fn(async (keys, callback = () => {}) => {
      return new Promise<{ [key: string]: string }>((resolve) => {
        console.log("get keys", keys);
        const items: { [key: string]: string } = {};
        if (typeof keys === "string") {
          items[keys] = storage[keys];
        } else if (Array.isArray(keys)) {
          keys.forEach((key) => {
            items[key] = storage[key];
          });
        } else if (keys === null) {
          Object.assign(items, storage);
        } else {
          Object.keys(keys).forEach((key) => {
            items[key] = storage[key];
          });
        }
        callback(items);
        resolve(items);
      });
    }),
    clear: jest.fn(() => {
      storage = {};
    }),
  };
})();

globalThis.chrome = {
  storage: {
    // @ts-ignore
    local: storageMock,
  },
};

import { act, renderHook } from "@testing-library/react";
import { useReferenceStore } from "../logic/reference_state";
import { Reference } from "../utils/message";

jest.mock("../utils/page_content", () => ({
  getCurrentPageRef: jest
    .fn()
    .mockImplementation(() => Promise.resolve(new Reference("webpage", "title1", "url1", "page1"))),
  getCurrentSelectionRef: jest
    .fn()
    .mockResolvedValue(() => new Reference("text", "title2", "url2", "text1")),
}));

describe("useReferenceStore", () => {
  beforeEach(() => {
    const { result } = renderHook(() => useReferenceStore());
    act(() => {
      result.current.clear();
    });
  });

  it("should add a page reference", async () => {
    const { result } = renderHook(() => useReferenceStore());
    expect(result.current.references).toEqual([]);

    await act(async () => {
      await result.current.addPageRef();
    });
    expect(result.current.references).toMatchSnapshot([{ title: "title1" }]);

    await act(async () => {
      await result.current.addSelectionRef();
    });
    expect(result.current.references).toMatchSnapshot([{ title: "title1" }, { title: "title2" }]);

    // should ignore duplicate
    await act(async () => {
      await result.current.addPageRef();
    });
    console.log("result1", result.current.references);
    expect(result.current.references).toMatchSnapshot([{ title: "title1" }, { title: "title2" }]);
  });

  it("should add a selection reference", async () => {
    const { result } = renderHook(() => useReferenceStore());
    console.log("result2", result.current.references);
    expect(result.current.references.length).toBe(0);
    await act(async () => {
      await result.current.addSelectionRef();
    });
    console.log("result3", result.current.references.length);

    expect(result.current.references.length).toBe(1);

    act(() => result.current.clear());
    expect(result.current.references.length).toBe(0);
  });

  it("should remove a reference", async () => {
    const { result } = renderHook(() => useReferenceStore());
    await act(async () => {
      await result.current.addPageRef();
      await result.current.addSelectionRef();
    });
    const referenceToRemove = result.current.references[0];
    act(() => {
      result.current.removeRef(referenceToRemove.id);
    });

    expect(result.current.references.length).toBe(1);
    expect(result.current.references[0].id).not.toBe(referenceToRemove.id);
  });

  it("should clear all references", () => {
    const { result } = renderHook(() => useReferenceStore());
    act(() => {
      result.current.addPageRef();
      result.current.addSelectionRef();
    });
    act(() => {
      result.current.clear();
    });

    expect(result.current.references.length).toBe(0);
  });

  // it("should load references from storage", async () => {
  // await (async () => {
  //   const { result } = renderHook(() => useReferenceStore());
  //   await act(async () => {
  //     await result.current.addPageRef();
  //   });
  // })();
  // const { result } = renderHook(() => useReferenceStore());
  // expect(result.current.references.length).toBe(0);
  // await act(async () => {
  //   await result.current.load();
  // });
  //   expect(result2.current.references.length).toBe(1);
  // });
});
