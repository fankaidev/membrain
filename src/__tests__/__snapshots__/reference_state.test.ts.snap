// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useReferenceStore should add a page reference 1`] = `
[
  {
    "content": "page1",
    "id": "webpage:url1",
    "title": "title1",
    "type": "webpage",
    "url": "url1",
  },
]
`;
